### PR TITLE
Filter writing tutorial. (Susan Sun)

### PR DESCRIPTION
**Pull request by SusanSun17**
## Description

A bounds check filter was added - header file PracticalBoundsCheck.h, implementation PracticalBoundsCheck.cc to src/ufo/filters, and they were added to the filter factory. The filter rejects observation values that fall outside of a specified bound. A test was written for the filter (test/testinput/qc_practical_boundscheck.yaml), and the appropriate CMakeLists.txt files were updated to include the test.

## Definition of Done

The test should pass/fail in appropriate situations.

### Issue(s) addressed

- fixes #<issue_number>

## Dependencies

This should not depend on any other PRs.

## Impact

No other repos will be affected and no updates are needed to test data.

### CODE REVIEW

File: test/testinput/qc_practical_boundscheck.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    minvalue: 14.0
+    maxvalue: 19.0
+#  Compare variables with minvalue/maxvalue
+#  variable1@ObsValue = 10, 11, 12, 13, 14, 15, 16, 17, 18, 19
+#  variable2@ObsValue = 10, 12, 14, 16, 18, 20, 22, 24, 26, 28
+#  variable3@ObsValue = 25, 24, 23, 22, 21, 20, 19, 18, 17, 16
+#  number of data points passing the filter
+  passedBenchmark: 13
+  
+observations:
```
Comment (twsearle) [inline test/testinput/qc_practical_boundscheck.yaml:27]: Not 100%, but I think you need to remove this observations key to run the second test?

File: test/testinput/qc_practical_boundscheck.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    - name: variable2
+    - name: variable3
+    minvalue: 15.0
+    maxvalue: 20.0
+#  Compare variables with minvalue/maxvalue
+#  variable1@ObsValue = 10, 11, 12, 13, 14, 15, 16, 17, 18, 19
+#  variable2@ObsValue = 10, 12, 14, 16, 18, 20, 22, 24, 26, 28
+#  variable3@ObsValue = 25, 24, 23, 22, 21, 20, 19, 18, 17, 16
+#  number of data points passing the filter
+  passedBenchmark: 18
```
Comment (twsearle) [inline test/testinput/qc_practical_boundscheck.yaml:47]: I think this might need to be 8, but again I am beginning to doubt myself.

**Review by twsearle — CHANGES_REQUESTED** [review]
Can you review which of us have made a mistake in our tests with me?

File: test/testinput/qc_practical_boundscheck.yaml (hunk 2 above)
Comment (SusanSun17) [reply test/testinput/qc_practical_boundscheck.yaml:47]: Thanks for your review! You picked up on my deliberate mistake! Yes, passedBenchmark should be 8 in that case - I will change it. I don't know about the 2nd 'observations' key - I thought the same as you, but it still runs the 2nd test with it like that.

Comment (twsearle) [reply test/testinput/qc_practical_boundscheck.yaml:47]: oh OK! all good then

File: test/testinput/qc_practical_boundscheck.yaml (hunk 1 above)
Comment (frld) [reply test/testinput/qc_practical_boundscheck.yaml:27]: Toby is right. I made the same mistake in my version which Toby is also reviewing. I think what happens with this there is that you have two things setting "observations" and the second one overwrites the first definition of observations.

Comment (twsearle) [reply test/testinput/qc_practical_boundscheck.yaml:27]: looks great!

Comment (SusanSun17) [reply test/testinput/qc_practical_boundscheck.yaml]: You're right @frld , I didn't realise until I forced the 1st test to fail and it didn't. I have corrected this now.

**Review by JorgeBornemann — APPROVED** [review]

**Pull request closed without merging**